From 31f03b6c24f471a17b4f74b31e93ec53aafa4566 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20K=C3=B6lbl?= <besentv@gmail.com>
Date: Sun, 30 Jan 2022 18:34:18 +0100
Subject: [PATCH 10/19] include/windows.media.speechrecognition.idl: Add
 SpeechRecognizerTimeouts.

---
 include/windows.media.speechrecognition.idl | 26 +++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/include/windows.media.speechrecognition.idl b/include/windows.media.speechrecognition.idl
index 76352531116..6ae1baf2040 100644
--- a/include/windows.media.speechrecognition.idl
+++ b/include/windows.media.speechrecognition.idl
@@ -46,6 +46,7 @@ namespace Windows {
             interface ISpeechRecognitionResult;
             interface ISpeechRecognitionResult2;
             interface ISpeechRecognitionSemanticInterpretation;
+            interface ISpeechRecognizerTimeouts;
             runtimeclass SpeechContinuousRecognitionCompletedEventArgs;
             runtimeclass SpeechContinuousRecognitionResultGeneratedEventArgs;
             runtimeclass SpeechContinuousRecognitionSession;
@@ -53,6 +54,7 @@ namespace Windows {
             runtimeclass SpeechRecognitionHypothesisGeneratedEventArgs;
             runtimeclass SpeechRecognitionResult;
             runtimeclass SpeechRecognitionSemanticInterpretation;
+            runtimeclass SpeechRecognizerTimeouts;
         }
     }
 }
@@ -242,6 +244,21 @@ namespace Windows {
                 [propget] HRESULT Properties([out, retval] Windows.Foundation.Collections.IMapView<HSTRING, Windows.Foundation.Collections.IVectorView<HSTRING>*> **value);
             }
 
+            [
+                contract(Windows.Foundation.UniversalApiContract, 1.0),
+                exclusiveto(Windows.Media.SpeechRecognition.SpeechRecognizerTimeouts),
+                uuid(2ef76fca-6a3c-4dca-a153-df1bc88a79af)
+            ]
+            interface ISpeechRecognizerTimeouts : IInspectable
+            {
+                [propget] HRESULT InitialSilenceTimeout([out, retval] Windows.Foundation.TimeSpan *value);
+                [propput] HRESULT InitialSilenceTimeout([in] Windows.Foundation.TimeSpan value);
+                [propget] HRESULT EndSilenceTimeout([out, retval] Windows.Foundation.TimeSpan *value);
+                [propput] HRESULT EndSilenceTimeout([in] Windows.Foundation.TimeSpan value);
+                [propget] HRESULT BabbleTimeout([out, retval] Windows.Foundation.TimeSpan *value);
+                [propput] HRESULT BabbleTimeout([in] Windows.Foundation.TimeSpan value);
+            }
+
             [
                 contract(Windows.Foundation.UniversalApiContract, 1.0),
                 marshaling_behavior(agile),
@@ -310,6 +327,15 @@ namespace Windows {
             {
                 [default] interface Windows.Media.SpeechRecognition.ISpeechRecognitionSemanticInterpretation;
             }
+
+            [
+                contract(Windows.Foundation.UniversalApiContract, 1.0),
+                marshaling_behavior(agile)
+            ]
+            runtimeclass SpeechRecognizerTimeouts
+            {
+                [default] interface Windows.Media.SpeechRecognition.ISpeechRecognizerTimeouts;
+            }
         }
     }
 }
-- 
2.34.1

