From a64e5030f0d5a977baba56f959f86ee90e2e4d23 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20K=C3=B6lbl?= <besentv@gmail.com>
Date: Sun, 30 Jan 2022 19:47:53 +0100
Subject: [PATCH v2 11/35] include/windows.media.speechrecognition.idl: Add
 SpeechRecognizerUIOptions.

---
 include/windows.media.speechrecognition.idl | 28 +++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/include/windows.media.speechrecognition.idl b/include/windows.media.speechrecognition.idl
index 6ae1baf2040..e2217e72528 100644
--- a/include/windows.media.speechrecognition.idl
+++ b/include/windows.media.speechrecognition.idl
@@ -47,6 +47,7 @@ namespace Windows {
             interface ISpeechRecognitionResult2;
             interface ISpeechRecognitionSemanticInterpretation;
             interface ISpeechRecognizerTimeouts;
+            interface ISpeechRecognizerUIOptions;
             runtimeclass SpeechContinuousRecognitionCompletedEventArgs;
             runtimeclass SpeechContinuousRecognitionResultGeneratedEventArgs;
             runtimeclass SpeechContinuousRecognitionSession;
@@ -55,6 +56,7 @@ namespace Windows {
             runtimeclass SpeechRecognitionResult;
             runtimeclass SpeechRecognitionSemanticInterpretation;
             runtimeclass SpeechRecognizerTimeouts;
+            runtimeclass SpeechRecognizerUIOptions;
         }
     }
 }
@@ -259,6 +261,23 @@ namespace Windows {
                 [propput] HRESULT BabbleTimeout([in] Windows.Foundation.TimeSpan value);
             }
 
+            [
+                contract(Windows.Foundation.UniversalApiContract, 1.0),
+                exclusiveto(Windows.Media.SpeechRecognition.SpeechRecognizerUIOptions),
+                uuid(7888d641-b92b-44ba-a25f-d1864630641f)
+            ]
+            interface ISpeechRecognizerUIOptions : IInspectable
+            {
+                [propget] HRESULT ExampleText([out, retval] HSTRING *value);
+                [propput] HRESULT ExampleText([in] HSTRING value);
+                [propget] HRESULT AudiblePrompt([out, retval] HSTRING *value);
+                [propput] HRESULT AudiblePrompt([in] HSTRING value);
+                [propget] HRESULT IsReadBackEnabled([out, retval] boolean *value);
+                [propput] HRESULT IsReadBackEnabled([in] boolean value);
+                [propget] HRESULT ShowConfirmation([out, retval] boolean *value);
+                [propput] HRESULT ShowConfirmation([in] boolean value);
+            }
+
             [
                 contract(Windows.Foundation.UniversalApiContract, 1.0),
                 marshaling_behavior(agile),
@@ -336,6 +355,15 @@ namespace Windows {
             {
                 [default] interface Windows.Media.SpeechRecognition.ISpeechRecognizerTimeouts;
             }
+
+            [
+                contract(Windows.Foundation.UniversalApiContract, 1.0),
+                marshaling_behavior(agile)
+            ]
+            runtimeclass SpeechRecognizerUIOptions
+            {
+                [default] interface Windows.Media.SpeechRecognition.ISpeechRecognizerUIOptions;
+            }
         }
     }
 }
-- 
2.34.1

