From 4f58cacd79757023c86e291912628bc7cb161d19 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20K=C3=B6lbl?= <besentv@gmail.com>
Date: Sun, 30 Jan 2022 20:16:39 +0100
Subject: [PATCH 12/24] include/windows.media.speechrecognition.idl: Add
 SpeechRecognitionCompilationResult.

---
 include/windows.media.speechrecognition.idl | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/include/windows.media.speechrecognition.idl b/include/windows.media.speechrecognition.idl
index e2217e72528..7d323d84b3a 100644
--- a/include/windows.media.speechrecognition.idl
+++ b/include/windows.media.speechrecognition.idl
@@ -40,6 +40,7 @@ namespace Windows {
             interface ISpeechContinuousRecognitionCompletedEventArgs;
             interface ISpeechContinuousRecognitionResultGeneratedEventArgs;
             interface ISpeechContinuousRecognitionSession;
+            interface ISpeechRecognitionCompilationResult;
             interface ISpeechRecognitionConstraint;
             interface ISpeechRecognitionHypothesis;
             interface ISpeechRecognitionHypothesisGeneratedEventArgs;
@@ -51,6 +52,7 @@ namespace Windows {
             runtimeclass SpeechContinuousRecognitionCompletedEventArgs;
             runtimeclass SpeechContinuousRecognitionResultGeneratedEventArgs;
             runtimeclass SpeechContinuousRecognitionSession;
+            runtimeclass SpeechRecognitionCompilationResult;
             runtimeclass SpeechRecognitionHypothesis;
             runtimeclass SpeechRecognitionHypothesisGeneratedEventArgs;
             runtimeclass SpeechRecognitionResult;
@@ -173,6 +175,16 @@ namespace Windows {
                 [eventremove] HRESULT ResultGenerated([in] EventRegistrationToken token);
             }
 
+            [
+                contract(Windows.Foundation.UniversalApiContract, 1.0),
+                exclusiveto(Windows.Media.SpeechRecognition.SpeechRecognitionCompilationResult),
+                uuid(407e6c5d-6ac7-4da4-9cc1-2fce32cf7489)
+            ]
+            interface ISpeechRecognitionCompilationResult : IInspectable
+            {
+                [propget] HRESULT Status([out, retval] Windows.Media.SpeechRecognition.SpeechRecognitionResultStatus *value);
+            }
+
             [
                 contract(Windows.Foundation.UniversalApiContract, 1.0),
                 uuid(79ac1628-4d68-43c4-8911-40dc4101b55b)
@@ -308,6 +320,15 @@ namespace Windows {
                 [default] interface Windows.Media.SpeechRecognition.ISpeechContinuousRecognitionSession;
             }
 
+            [
+                contract(Windows.Foundation.UniversalApiContract, 1.0),
+                marshaling_behavior(agile)
+            ]
+            runtimeclass SpeechRecognitionCompilationResult
+            {
+                [default] interface Windows.Media.SpeechRecognition.ISpeechRecognitionCompilationResult;
+            }
+
             [
                 contract(Windows.Foundation.UniversalApiContract, 1.0),
                 marshaling_behavior(agile),
-- 
2.34.1

