From 9b2706f4f5c1fab0787b2c68fde4f28d1f5b5fb5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20K=C3=B6lbl?= <besentv@gmail.com>
Date: Fri, 4 Feb 2022 22:42:39 +0100
Subject: [PATCH v2 37/41] windows.media.speech: Implement IVector<HSTRING>
 Append.

---
 .../speechrecognitionlistconstraint.c              | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/dlls/windows.media.speech/speechrecognitionlistconstraint.c b/dlls/windows.media.speech/speechrecognitionlistconstraint.c
index 62d71d71ddc..36908dd2c7c 100644
--- a/dlls/windows.media.speech/speechrecognitionlistconstraint.c
+++ b/dlls/windows.media.speech/speechrecognitionlistconstraint.c
@@ -161,8 +161,18 @@ static HRESULT STDMETHODCALLTYPE vector_hstring_RemoveAt(
 static HRESULT STDMETHODCALLTYPE vector_hstring_Append(
     IVector_HSTRING *iface, HSTRING value)
 {
-    FIXME("iface %p, value %p, stub!\n", iface, value);
-    return E_NOTIMPL;
+    struct vector_hstring *impl = impl_from_IVector_HSTRING(iface);
+
+    TRACE("iface %p, value %s, stub!\n", iface, debugstr_hstring(value));
+
+    if(!value)
+        return E_POINTER;
+
+    impl->count++;
+    impl->values = realloc(impl->values, impl->count * sizeof(HSTRING));
+    impl->values[impl->count-1] = value;
+
+    return S_OK;
 }
 
 static HRESULT STDMETHODCALLTYPE vector_hstring_RemoveAtEnd(
-- 
2.34.1

