From a88f232bbb420048b9d62b567734d4021f481812 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20K=C3=B6lbl?= <besentv@gmail.com>
Date: Fri, 4 Feb 2022 23:28:21 +0100
Subject: [PATCH v2 38/45] windows.media.speech: Store voice commands in a
 vector.

---
 .../speechrecognitionlistconstraint.c          | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/dlls/windows.media.speech/speechrecognitionlistconstraint.c b/dlls/windows.media.speech/speechrecognitionlistconstraint.c
index 36908dd2c7c..1d328661a7f 100644
--- a/dlls/windows.media.speech/speechrecognitionlistconstraint.c
+++ b/dlls/windows.media.speech/speechrecognitionlistconstraint.c
@@ -535,6 +535,7 @@ static HRESULT STDMETHODCALLTYPE speech_recognition_list_constraint_create_with_
     IIterable_HSTRING *commands, HSTRING tag, ISpeechRecognitionListConstraint **listconstraint)
 {
     struct speech_recognition_list_constraint *impl;
+    IIterator_HSTRING *iterator;
     HRESULT hr;
 
     FIXME("commands %p, tag %p, listconstraint %p. Stub!\n", commands, tag, listconstraint);
@@ -554,7 +555,22 @@ static HRESULT STDMETHODCALLTYPE speech_recognition_list_constraint_create_with_
 
     impl->enabled = TRUE;
     hr = vector_hstring_create(NULL, 0, &impl->commands);
-    
+    hr = IIterable_HSTRING_First(commands, &iterator);
+    if(hr == S_OK)
+    {
+        BOOL value_available;
+        HSTRING value;
+
+        IIterator_HSTRING_get_HasCurrent(iterator, &value_available);
+        while(value_available)
+        {
+            IIterator_HSTRING_get_Current(iterator, &value);
+            IVector_HSTRING_Append(impl->commands, value);
+            IIterator_HSTRING_MoveNext(iterator, &value_available);
+        }
+        IIterator_HSTRING_Release(iterator);
+    }
+
     hr = ISpeechRecognitionListConstraint_QueryInterface(&impl->ISpeechRecognitionListConstraint_iface, &IID_ISpeechRecognitionListConstraint, (void**)listconstraint);
     ISpeechRecognitionListConstraint_Release(&impl->ISpeechRecognitionListConstraint_iface);
 
-- 
2.34.1

